feat(package_info_plus)!: Add information about the installer store

ueman opened the pull request:
## Description

Adds the store through which the application was installed to the package info class.
This currently supports Android, iOS and macOS.

On web, it could make sense to return the URL, but I'm not too sure about that. Let me know what you think.

I'm also not sure what to do with Linux and Windows. It might be possible but I don't have a Windows machine nor a linux machine to test and figure it out. For now, both return just `null` for the installer store.

I'm not sure whether this is considered a breaking change, so I'd like to hear from you guys. Edit: It's a breaking change now.

## Related Issues

Fixes https://github.com/fluttercommunity/plus_plugins/issues/570

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated the version in `pubspec.yaml` and `CHANGELOG.md`.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/fluttercommunity/plus_plugins/blob/main/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

miquelbeltran commented:
I think it is not completely a breaking change, but since you are changing the platform interface now all versions need to be bumped to major version so dependencies are resolved correctly.

ueman commented:
> I think it is not completely a breaking change, but since you are changing the platform interface now all versions need to be bumped to major version so dependencies are resolved correctly.

Thanks, will do so 👍

ueman commented on packages/package_info_plus/package_info_plus_web/lib/package_info_plus_web.dart:
```diff
@@ -43,6 +43,7 @@ class PackageInfoPlugin extends PackageInfoPlatform {
       buildNumber: versionMap['build_number'] ?? '',
       packageName: versionMap['package_name'] ?? '',
       buildSignature: '',
+      installerStore: null,
```
I'm wondering whether it makes sense to return the current URL here

miquelbeltran replied:
I think null is fine

vbuberen replied:
Agree that `null` is fine here.

ueman replied:
Gotcha, `null` it is.

miquelbeltran commented:
Could you update the integration_test as well to include this new property?

I expect it will be null always when running integration_tests, right?

vbuberen commented on packages/package_info_plus/package_info_plus_platform_interface/lib/package_info_data.dart:
```diff
@@ -9,6 +9,7 @@ class PackageInfoData {
     required this.version,
     required this.buildNumber,
     required this.buildSignature,
+    required this.installerStore,
```
Why do we need to make it `required`, considering that we return this parameter only on Android, iOS and MacOS platforms? I would go with optional parameter instead.

ueman replied:
I have no special reason for making it required. I'll make it optional then.

vbuberen replied:
Awesome, thanks. With `required` it feels misleading for Linux, Web and Windows users to have this new parameter being added, but at the same time being not usable for them.

vbuberen reviewed (changes requested):
Had a thorough look at the PR and don't agree that we need to mark the new parameter as `required`, because we return proper value only on Android, iOS and MacOS platforms.

So disregard my previous comment about `null` being fine.

ueman commented:
> Could you update the integration_test as well to include this new property?

Yes, sure.

> I expect it will be null always when running integration_tests, right?

I'll figure it out and you'll be going to see when I update this PR :D

ueman commented:
After some more investigation, the implementation on macOS made no sense and I changed it to just return `null`. I've also squashed all commits into one and addressed the remaining feedback.

miquelbeltran reviewed (approved):
Thank you!

nohli commented on packages/package_info_plus/package_info_plus_macos/pubspec.yaml:
```diff
@@ -16,5 +16,6 @@ environment:
   flutter: ">=1.20.0"
 
 dependencies:
+  package_info_plus_platform_interface: ^2.0.0
```
Is there still a reason to add this dependency here?

ueman replied:
No, it doesn't seem so. I'll remove it

ueman replied:
```suggestion
```

nohli commented on packages/package_info_plus/package_info_plus_macos/pubspec.yaml:
```diff
@@ -1,6 +1,6 @@
 name: package_info_plus_macos
 description: macOS implementation of the package_info_plus plugin.
-version: 1.3.0
+version: 2.0.0
```
Does this need to be a breaking change?

ueman replied:
I wondered that too, and got this answer: https://github.com/fluttercommunity/plus_plugins/pull/1135#issuecomment-1264632986

nohli replied:
Ok, because of the federated plugin structure it's safer to update everything.

nohli reviewed (commented):
Do we need a breaking change in `package_info_plus_platform_interface` (and all other packages)?

ueman commented on packages/package_info_plus/package_info_plus_windows/CHANGELOG.md:
```diff
@@ -1,6 +1,6 @@
+## 3.0.0
 
-- Allow win32 3.x to be used.
+- Add information about the store, through which the application was installed
```
```suggestion
- Add information about the store, through which the application was installed

## 2.1.0

- Allow win32 3.x to be used.
```

vbuberen commented:
> After some more investigation, the implementation on macOS made no sense and I changed it to just return `null`. I've also squashed all commits into one and addressed the remaining feedback.

During my review I suggested to make the parameter optional and you did it, but I don't get it why still return `null` on platforms which don't provide the info about the installer. If the parameter is optional in the platform interface just skip it on platforms which don't have the info.

vbuberen reviewed (commented):
From my side everything looks good except that small change to not return `null` on unsupported platforms as I suggested in my comment above.

miquelbeltran commented on packages/package_info_plus/package_info_plus_linux/lib/src/package_info.dart:
```diff
@@ -23,6 +23,7 @@ class PackageInfoLinux extends PackageInfoPlatform {
       buildNumber: versionJson['build_number'] ?? '',
       packageName: versionJson['package_name'] ?? '',
       buildSignature: '',
+      installerStore: null,
```
What @vbuberen means is to not pass null to these parameters, as it is the default value.

(shouldn't that be picked by flutter analyze?)

ueman replied:
I got that, I'm just away from a computer until the end of the week, so I can't fix it until then.

nohli replied:
If my four suggestions below are ok, you (or I) can commit them via browser 🙂

miquelbeltran commented:
Hey! Due to changes in our contributor guidelines, all PRs should now:

- Have a title that follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/) format. e.g. `feat(package_name)!: feature description`
- **DO NOT** modify the CHANGELOG.md or the version in the `pubspec.yaml`. This is going to be an automated process now.

We would ask you kindly to update the PR following these changes.

Thanks!

vbuberen commented:
I will take care of updating this PR later today.

miquelbeltran commented:
> I will take care of updating this PR later today.

Assigned it to you, so we remember :)

vbuberen commented:
Looks like @ueman returned earlier when I :)

vbuberen reviewed (approved):
Looks good to me now. Thanks a lot for your contribution @ueman 

I will merge as soon as tests finish.

ueman commented:
> Looks good to me now. Thanks a lot for your contribution @ueman
> 
> I will merge as soon as tests finish.

It was a pleasure

ekuleshov-idexx commented:
> After some more investigation, the implementation on macOS made no sense and I changed it to just return `null`. I've also squashed all commits into one and addressed the remaining feedback.

Thank you for adding this.

Though I'm curious what was the issue with macos? Mac apps can be distributed through testflight or the app store and the docs for `appStoreReceiptURL` say it is supported on MacOS 10.7+

https://developer.apple.com/documentation/foundation/nsbundle/1407276-appstorereceipturl?language=objc

ueman commented:
> > After some more investigation, the implementation on macOS made no sense and I changed it to just return `null`. I've also squashed all commits into one and addressed the remaining feedback.
> 
> Thank you for adding this.
> 
> Though I'm curious what was the issue with macos? Mac apps can be distributed through testflight or the app store and the docs for `appStoreReceiptURL` say it is supported on MacOS 10.7+
> 
> https://developer.apple.com/documentation/foundation/nsbundle/1407276-appstorereceipturl?language=objc

Yeah, but the path doesn't follow the same schema, so the same trick as in iOS doesn't work. Feel free to improve the code though, if you find a way to get some information out of it.

calcitem commented:
Can the README be updated? I still don't know how to use this feature? Thanks!

dgilperez commented:
Guys just coming to THANK YOU for this PR. This is going to hugely help our growth and analytics team and simplify our testing processes. Kudos!
